### PR TITLE
Add `LightGBM`

### PR DESCRIPTION
**Pull request by xnuohz**
From https://github.com/pyg-team/pytorch-frame/issues/245
### Benchmark

- Binary Classification + Small + 0
```
Best is trial 17 with value: 0.9283801711103639.
{'args': {'task_type': 'binary_classification', 'scale': 'small', 'idx': 0, 'epochs': 50, 'num_trials': 20, 'num_repeats': 5, 'model_type': 'LightGBM', 'seed': 0, 'result_path': './output/results.pt'}, 'best_val_metric': 0.9283801711103639, 'best_test_metric': 0.9314460189332364, 'best_cfg': {'verbosity': -1, 'bagging_freq': 1, 'max_depth': 8, 'learning_rate': 0.009535731905301808, 'num_leaves': 428, 'subsample': 0.7982234374896812, 'colsample_bytree': 0.4258609181729451, 'reg_alpha': 0.19993496122846377, 'reg_lambda': 0.0117270935126881, 'min_child_samples': 64, 'min_data_in_leaf': 64, 'objective': 'binary', 'metric': 'auc'}, 'total_time': 21.042486667633057}
```
- Regression + Small + 0

```
Best is trial 18 with value: 0.21210491971461293.
{'args': {'task_type': 'regression', 'scale': 'small', 'idx': 0, 'epochs': 50, 'num_trials': 20, 'num_repeats': 5, 'model_type': 'LightGBM', 'seed': 0, 'result_path': './output/results.pt'}, 'best_val_metric': 0.21210491971461293, 'best_test_metric': 0.24810077896674687, 'best_cfg': {'verbosity': -1, 'bagging_freq': 1, 'max_depth': 8, 'learning_rate': 0.012302993351171673, 'num_leaves': 1022, 'subsample': 0.8246554962306221, 'colsample_bytree': 0.7730036635179613, 'reg_alpha': 0.0001244683549932226, 'reg_lambda': 8.8690473810824e-05, 'min_child_samples': 33, 'min_data_in_leaf': 13, 'objective': 'regression', 'metric': 'rmse'}, 'total_time': 25.504676342010498}
```

- Multiclass Classification + medium + 0

```Too slow*```

### CODE REVIEW

File: torch_frame/gbdt/tuned_lightgbm.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            trial (optuna.trial.Trial): Optuna trial object.
+            tf_train (TensorFrame): Train data.
+            tf_val (TensorFrame): Validation data.
+            num_boost_round (int): Number of boosting round.
+
+        Returns:
+            float: Best objective value. Mean absolute error for
+            regression task and accuracy for classification task.
+        """
+        import lightgbm
```
Comment (yiweny) [inline torch_frame/gbdt/tuned_lightgbm.py:100]: You need to add the dependency of lightgbm in `pyproject.toml`

Comment (xnuohz) [reply torch_frame/gbdt/tuned_lightgbm.py:101]: Fixed.

File: torch_frame/gbdt/tuned_lightgbm.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+            feats.append(tf.feat_dict[stype.categorical])
+
+        if stype.numerical in tf.feat_dict:
+            feats.append(tf.feat_dict[stype.numerical])
+
+        if stype.text_embedded in tf.feat_dict:
+            feat = tf.feat_dict[stype.text_embedded]
+            feat = feat.values
+            feat = feat.view(feat.size(0), -1)
+            feats.append(feat)
```
Comment (weihua916) [inline torch_frame/gbdt/tuned_lightgbm.py:50]: How does LightGBM distinguish numerical and categorical features?

File: benchmark/README.md [hunk 3]
```diff
@@ -135,12 +136,12 @@ Experimental setting: 20 Optuna search trials for XGBoost and CatBoost. 5 Optuna
...
-| FTTransformerBucket | **0.879±0.006 (9104s)** | 0.365±0.002 (1067s)   | 0.722±0.001 (2366s)    |
-| ExcelFormer         | OOM                     | 0.375±0.004 (2168s)   | 0.732±0.000 (4138s)    |
+| ResNet              | **0.951±0.000 (419s)**  | **0.378±0.001 (171s)**| 0.723±0.001 (257s)     |
+| FTTransformerBucket | 0.879±0.006 (9104s)     | 0.365±0.002 (1067s)   | 0.722±0.001 (2366s)    |
+| ExcelFormer         | OOM                     | 0.375±0.004 (2168s)   | **0.732±0.000 (4138s)**|
 | FTTransformer       | 0.923±0.003 (14517s)    | 0.357±0.001 (754s)    | 0.724±0.004 (2621s)    |
-| TabNet              | 0.934±0.005 (218s)      | **0.349±0.001 (64s)** | 0.716±0.001 (153s)     |
-| TabTransformer      | 0.950±0.000 (160s)      | 0.352±0.001 (98s)     | **0.705±0.003 (103s)** |
+| TabNet              | 0.934±0.005 (218s)      | 0.349±0.001 (64s)     | 0.716±0.001 (153s)     |
+| TabTransformer      | 0.950±0.000 (160s)      | 0.352±0.001 (98s)     | 0.705±0.003 (103s)     |
```
Comment (weihua916) [inline benchmark/README.md:144]: Oops, thanks :)

File: torch_frame/gbdt/tuned_lightgbm.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                self.params["metric"] = "rmse"
+            elif self.metric == Metric.MAE:
+                self.params["objective"] = "regression_l1"
+                self.params["metric"] = "mae"
+        elif self.task_type == TaskType.BINARY_CLASSIFICATION:
+            self.params["objective"] = "binary"
+            if self.metric == Metric.ROCAUC:
+                self.params["metric"] = "auc"
+            elif self.metric == Metric.ACCURACY:
+                self.params["metric"] = "binary_logloss"
```
Comment (weihua916) [inline torch_frame/gbdt/tuned_lightgbm.py:140]: Doesn't LightGBM offer accuracy?

**Review by weihua916 — COMMENTED** [review]
Thanks! very nice. Left some comments to address.

File: torch_frame/gbdt/tuned_lightgbm.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+import numpy as np
+import torch
+from torch import Tensor
+
+from torch_frame import DataFrame, Metric, TaskType, TensorFrame, stype
+from torch_frame.gbdt import GBDT
+
+
+class LightGBM(GBDT):
+    r"""A LightGBM model implementation with hyper-parameter tuning using
```
Comment (yiweny) [inline torch_frame/gbdt/tuned_lightgbm.py:12]:
```suggestion
    r"""LightGBM implementation with hyper-parameter tuning using
```

File: torch_frame/gbdt/tuned_lightgbm.py (hunk 4 above)
Comment (xnuohz) [reply torch_frame/gbdt/tuned_lightgbm.py:152]: https://lightgbm.readthedocs.io/en/latest/Parameters.html#metric

File: torch_frame/gbdt/tuned_lightgbm.py (hunk 5 above)
Comment (xnuohz) [reply torch_frame/gbdt/tuned_lightgbm.py]: Updated.

File: torch_frame/gbdt/tuned_lightgbm.py (hunk 4 above)
Comment (weihua916) [reply torch_frame/gbdt/tuned_lightgbm.py:152]: I think binary error is what you should use. It’s equivalent to accuracy

**Review by weihua916 — APPROVED** [review]
LGTM!

**Review by yiweny — APPROVED** [review]

File: torch_frame/gbdt/tuned_lightgbm.py (hunk 2 above)
Comment (weihua916) [reply torch_frame/gbdt/tuned_lightgbm.py]: Thanks!
